File: zen-rhino/src/test/resources/com/nominanuda/rhino/RhinoOverloadingTest.js
const CONSOLE = require('console');
const JUNIT = require('classpath:com/nominanuda/rhino/JUnit.js');


function test(v, r, t) {
	if (t) {
		JUNIT.assert(r === t);
		CONSOLE.log(v + ' as ' + r + ' OK!');
	} else {
		CONSOLE.log(v + ' as ' + r + ' ???');
	}
}


var o = require('overloads');
CONSOLE.log('\n\n overloads');
test(1, o.method(1), 'number');
test('s', o.method('s'), 'string');
test([], o.method([]), 'json');
test({}, o.method({}), 'json');


var o = require('numberOverload');
CONSOLE.log('\n\n number overload');
test(1, o.method(1), 'number');
test('s', o.method('s'), 'object');
test([], o.method([]), 'object');
test({}, o.method({}), 'object');


var o = require('stringOverload');
CONSOLE.log('\n\n string overload');
test(1, o.method(1));
test('s', o.method('s'), 'string');
test([], o.method([]), 'object');
test({}, o.method({}), 'object');


var o = require('arrayOverload');
CONSOLE.log('\n\n array overload');
test(1, o.method(1), 'object');
test('s', o.method('s'), 'object');
test([], o.method([]), 'array');
test({}, o.method({}), 'object');


var o = require('mapOverload');
CONSOLE.log('\n\n map overload');
test(1, o.method(1), 'object');
test('s', o.method('s'), 'object');
test([], o.method([]), 'object');
test({}, o.method({}), 'map');


var o = require('jsonOverload');
CONSOLE.log('\n\n json overload');
test(1, o.method(1), 'object');
test('s', o.method('s'), 'object');
test([], o.method([]), 'json');
test({}, o.method({}), 'json');


var o = require('noOverloads');
CONSOLE.log('\n\n no overloads');
test(1, o.method(1), 'object');
test('s', o.method('s'), 'object');
test([], o.method([]), 'object');
test({}, o.method({}), 'object');


var o = require('instanceOf');
CONSOLE.log('\n\n instance of');
test(1, o.method(1), 'object');
test('s', o.method('s'), 'object');
test([], o.method([]), 'array');
test({}, o.method({}), 'map');

File: zen-rhino/src/test/java/com/nominanuda/rhino/RhinoOverloadingTest.java
package com.nominanuda.rhino;

import java.util.Map;

import com.nominanuda.zen.obj.Arr;
import com.nominanuda.zen.obj.Obj;
import com.nominanuda.zen.obj.Stru;

public class RhinoOverloadingTest extends RhinoTestCase {
	@Override
	@SuppressWarnings("unused")
	protected Map<String, Object> buildJavaObjectsMap() {
		Map<String, Object> m = super.buildJavaObjectsMap();
		
		m.put("overloads", new Object() {
			public String method(Number n) {
				return "number";
			}
			public String method(String s) {
				return "string";
			}
			public String method(Stru j) {
				return "json";
			}
		});
		
		m.put("numberOverload", new Object() {
			public String method(Number n) {
				return "number";
			}
			public String method(Object o) {
				return "object";
			}
		});
		
		m.put("stringOverload", new Object() {
			public String method(String s) {
				return "string";
			}
			public String method(Object o) {
				return "object";
			}
		});
		
		m.put("arrayOverload", new Object() {
			public String method(Arr a) {
				return "array";
			}
			public String method(Object o) {
				return "object";
			}
		});
		
		m.put("mapOverload", new Object() {
			public String method(Obj a) {
				return "map";
			}
			public String method(Object o) {
				return "object";
			}
		});
		
		m.put("jsonOverload", new Object() {
			public String method(Stru j) {
				return "json";
			}
			public String method(Object o) {
				return "object";
			}
		});
		
		m.put("noOverloads", new Object() {
			public String method(Object o) {
				return "object";
			}
		});
		
		m.put("instanceOf", new Object() {
			public String method(Object o) {
				if(o instanceof Arr) {
					return "array";
				}
				if(o instanceof Obj) {
					return "map";
				}
				return "object";
			}
		});
		
		return m;
	}
}
